File: clojure/geometry.clj
(ns geometry)


(defrecord Point [^Number x, ^Number y])


(defn circle
  ([^Number radius]
   (circle (->Point 0 0) radius))
  ([^Point origin, ^Number radius]
   (fn [^Point p]
     (let [o_x (:x origin)
           o_y (:y origin)
           x (:x p)
           y (:y p)
           dx (- o_x x)
           dy (- o_y y)]
       (>= radius (Math/hypot dx dy))))))


(defn rectangle [^Point bottom-left, ^Number width, ^Number height]
  (fn [^Point p]
    (let [max-x (+ (:x bottom-left) width)
          min-x (:x bottom-left)
          max-y (+ (:y bottom-left) height)
          min-y (:y bottom-left)]
      (and (>= max-x (:x p) min-x)
           (>= max-y (:y p) min-y)))))


(defn square [^Point bottom-left, ^Number side-length]
  (rectangle bottom-left side-length side-length))


(defn or [& shapes]
  (fn [^Point p]
    (reduce (fn [shape]
              (if (shape p)
                (reduced true)
                false))
            false
            shapes)    ))


(defn and [& shapes]
  (fn [^Point p]
    (reduce (fn [shape]
              (if (not (shape p))
                (reduced false)
                true))
            true
            shapes)))

File: fsharp/geometry.fs
module Geometry

open System


type Point<'a> = { x: 'a; y: 'a }


type Shape<'a> = Point<'a> -> bool


let circle origin radius p =
    let dx = origin.x - p.x |> float
    let dy = origin.y - p.y |> float
    let hypot = dx*dx + dy*dy |> Math.Sqrt
    hypot <= radius


let rectangle lower_left upper_right p =
    ((lower_left.x <= p.x) &&
     (p.x <= upper_right.x) &&
     (lower_left.y <= p.y) &&
     (p.y <= upper_right.y))


let square lower_left side_length =
    rectangle lower_left { x=lower_left.x+side_length;
                           y=lower_left.y+side_length }


let translate shape difference p =
    shape { x=p.x-difference.x; y=p.y-difference.y }


let union shape_1 shape_2 p =
    shape_1 p || shape_2 p
